### PR TITLE
Instance rendering support

### PR DESCRIPTION
**Pull request by seandillon92**
This PR introduces support for rendering instances in the Unity Server.

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/unity3d-jp/MeshSync?pullRequest=540) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/unity3d-jp/MeshSync?pullRequest=540) before we can accept your contribution.<br/><hr/>**Sean Dillon** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/unity3d-jp/MeshSync?pullRequest=540) it.</sub>

File: Runtime/Scripts/BaseMeshSync.cs [hunk 1]
```diff
@@ -2232,8 +2232,27 @@ enum BaseMeshSyncVersion {
...
+        var renderer = GetComponent<MeshSyncInstanceRenderer>();
+        
+        if (active && renderer == null)
+        {
+            renderer = gameObject.AddComponent<MeshSyncInstanceRenderer>();
+            renderer.Init(this);
+        }
+        else if (!active && renderer != null)
+        {
+            #if UNITY_EDITOR
```
Comment (sindharta) [inline Runtime/Scripts/BaseMeshSync.cs]: I'd suggest using `FilmInternalUtilities.ObjectUtility.Destroy() ` to handle the branching.

File: Runtime/Scripts/Settings/Data/MeshSyncPlayerConfig.cs [hunk 2]
```diff
@@ -102,6 +103,8 @@ internal abstract class MeshSyncPlayerConfig : ISerializationCallbackReceiver {
     //Sync Settings
     public bool SyncVisibility = true;
     public bool SyncTransform  = true; //for Update
+
+    public bool SyncInstances = false;
```
Comment (sindharta) [inline Runtime/Scripts/Settings/Data/MeshSyncPlayerConfig.cs]: This should be in `MeshSyncServerConfig`, unless we need to support this for SceneCache as well

File: Runtime/Scripts/MeshSyncInstanceRenderer.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            }
+
+            if (!obj.TryGetComponent(out MeshRenderer renderer))
+            {
+                Debug.LogWarningFormat("Object {0} has instance but no MeshRenderer", obj.name);
+            }
+            
+            var mesh = meshFilter.sharedMesh;
+
+            var entry = meshInstances.Find(x => x.Mesh == mesh);
```
Comment (sindharta) [inline Runtime/Scripts/MeshSyncInstanceRenderer.cs]: Using HashSet or Dictionary might be better for lookup ?

File: Runtime/Scripts/MeshSyncInstanceRenderer.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+            if (entry.Materials.Length == 0)
+                return;
+            
+            for (var i = 0; i < mesh.subMeshCount; i++)
+            {
+                // Try to get the material in the same index position as the mesh
+                // or the last material.
+                
+                var materialIndex = Mathf.Max(Mathf.Min(entry.Materials.Length -1, i), 0);
```
Comment (sindharta) [inline Runtime/Scripts/MeshSyncInstanceRenderer.cs]: Isn't this the same as `Mathf.Clamp` ?

File: Runtime/Scripts/Settings/Data/MeshSyncPlayerConfig.cs (hunk 2 above)
Comment (seandillon92) [reply Runtime/Scripts/Settings/Data/MeshSyncPlayerConfig.cs]: Oops Good catch! I removed the toggle I introduced completely, instance rendering should be active by default now

File: Runtime/Scripts/BaseMeshSync.cs (hunk 1 above)
Comment (seandillon92) [reply Runtime/Scripts/BaseMeshSync.cs]: Oh nice, changed it to use the utility

File: Runtime/Scripts/MeshSyncInstanceRenderer.cs (hunk 3 above)
Comment (seandillon92) [reply Runtime/Scripts/MeshSyncInstanceRenderer.cs]: Good call, this is done now

File: Runtime/Scripts/MeshSyncInstanceRenderer.cs (hunk 4 above)
Comment (seandillon92) [reply Runtime/Scripts/MeshSyncInstanceRenderer.cs]: It is indeed, I changed it use the clamping function now

File: Editor/Scripts/ProjectSettings/MeshSyncPlayerConfigSection.cs [hunk 5]
```diff
@@ -17,6 +17,7 @@ internal class Contents {
 
         public static readonly GUIContent Visibility      = EditorGUIUtility.TrTextContent("Visibility");
         public static readonly GUIContent UpdateTransform = EditorGUIUtility.TrTextContent("Update Transform");
+        public static readonly GUIContent RenderInstances = EditorGUIUtility.TrTextContent("Render Instances");
```
Comment (seandillon92) [inline Editor/Scripts/ProjectSettings/MeshSyncPlayerConfigSection.cs:20]: will remove this

File: Plugin~/Src/MeshSync/Include/MeshSync/SceneGraph/msInstanceInfo.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#pragma once
+#include "MeshUtils/muMath.h"
+#include "MeshSync/msFoundation.h"
+
+namespace ms {
+	using namespace std;
+	using namespace mu;
+	class InstanceInfo
+	{
+		uint64_t m_hash = 0;
```
Comment (seandillon92) [inline Plugin~/Src/MeshSync/Include/MeshSync/SceneGraph/msInstanceInfo.h:12]: not sure what would be a good and efficient hash for this.. maybe a random number when the transforms member is assigned?

File: Plugin~/Src/mscore/msCoreAPI.cpp [hunk 7]
```diff
@@ -543,13 +543,22 @@ msAPI quatf msParentConstraintGetRotationOffset(const ms::ParentConstraint *self
...
 msAPI bool msSceneSubmeshesHaveUniqueMaterial(const ms::Scene *self) { return self->submeshesHaveUniqueMaterial(); }
 msAPI ms::SceneProfileData msSceneGetProfileData(const ms::Scene *self) { return self->profile_data; }
 #pragma endregion
 
+#pragma region InstanceInfo
+msAPI const char* msInstanceInfoGetPath(const ms::InstanceInfo* self) { return self->path.c_str(); }
+msAPI int msInstanceInfoPropGetArrayLength(const ms::InstanceInfo* self) { return self->transforms.size(); }
+msAPI void msInstanceInfoCopyTransforms(const ms::InstanceInfo* self, float4x4* dst) {
+    memcpy(dst, self->transforms.data(), self->transforms.size() * sizeof(float4x4));
+}
```
Comment (seandillon92) [inline Plugin~/Src/mscore/msCoreAPI.cpp:563]: I would prefer to avoid the copy here, but I am not 100% what happens with the ownership of the array

File: Runtime/Scripts/BaseMeshSync.cs [hunk 8]
```diff
@@ -1764,6 +1810,22 @@ void UpdateConstraint(ConstraintData data)
...
+    {
+        var path = identifier.name;
+
+        var ret = false;
+        if (this.m_clientInstances.TryGetValue(path, out InstanceInfoRecord rec))
+        {
+             ret = this.m_clientInstances.Remove(path);
+             
+            if (onDeleteInstanceInfo != null)
+                onDeleteInstanceInfo(rec.go);
```
Comment (seandillon92) [inline Runtime/Scripts/BaseMeshSync.cs]: Need to hook up the InstanceRenderer to listen to this event

Comment (seandillon92) [reply Runtime/Scripts/BaseMeshSync.cs]: Same for the onDeleteEntity event

File: Runtime/Scripts/PlayerData/InstanceInfoRecord.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+﻿using UnityEngine;
+
+namespace Unity.MeshSync
+{
+    internal class InstanceInfoRecord
+    {
+        public GameObject go;
+    }
```
Comment (seandillon92) [inline Runtime/Scripts/PlayerData/InstanceInfoRecord.cs]: This is WIP, for the moment I wanted to mimic how instances are managed. Am I right to think that the EntityRecord related to the CachePlayer? could we use the InstanceInfoRecord as a way to store instancing info on the cache?

File: Runtime/Scripts/Settings/Data/MeshSyncServerConfig.cs [hunk 10]
```diff
@@ -6,7 +6,10 @@ namespace Unity.MeshSync {
 internal class MeshSyncServerConfig : MeshSyncPlayerConfig {
     internal MeshSyncServerConfig() { }
 
-    internal MeshSyncServerConfig(MeshSyncServerConfig other) : base(other) { }
+    internal MeshSyncServerConfig(MeshSyncServerConfig other) : base(other)
+    {
+        SyncInstances        = other.SyncInstances;
+    }
```
Comment (seandillon92) [inline Runtime/Scripts/Settings/Data/MeshSyncServerConfig.cs]: Need to remove this change

File: Runtime/Scripts/Settings/Data/MeshSyncServerConfig.cs [hunk 11]
```diff
@@ -19,6 +22,8 @@ internal class MeshSyncServerConfig : MeshSyncPlayerConfig {
     enum MeshSyncServerConfigVersion {
         INITIAL = 1,
     }
+    
+    public bool SyncInstances = false;
```
Comment (seandillon92) [inline Runtime/Scripts/Settings/Data/MeshSyncServerConfig.cs]: Need to remove this change

File: Runtime/Scripts/msCoreAPI.cs [hunk 12]
```diff
@@ -1366,15 +1366,15 @@ public enum Type {
         Float4x4,
     }
 
-    string name {
+    internal string name {
```
Comment (seandillon92) [inline Runtime/Scripts/msCoreAPI.cs:1374]: These changes are not relevant anymore, need to revert them

File: Runtime/Scripts/msCoreAPI.cs [hunk 13]
```diff
@@ -2901,6 +2901,38 @@ internal struct SceneProfileData {
...
+
+    public int arrayLength => msInstanceInfoPropGetArrayLength(self);
+
+    public Matrix4x4[] transforms
+    {
+        get
+        {
+            var mat = new Matrix4x4[arrayLength];
+            msInstanceInfoCopyTransforms(self, mat);
+            return mat;
```
Comment (seandillon92) [inline Runtime/Scripts/msCoreAPI.cs]: will try to avoid the copy here

File: Runtime/Scripts/msMessagesAPI.cs [hunk 14]
```diff
@@ -82,6 +84,16 @@ internal struct DeleteMessage {
...
+    
+    public int numInstanceInfos
+    {
+        get { return msDeleteGetNumInstanceInfos((self)); }
+    }
+
+    internal Identifier GetInstanceInfo(int i)
+    {
+        return msDeleteGetInstanceInfo(self, i);
+    }
```
Comment (seandillon92) [inline Runtime/Scripts/msMessagesAPI.cs]: This is WIP, the entry points are not implemented yet

**Review by seandillon92 — COMMENTED** [review]
Hey @sindharta-tanuwijaya , started making the changes here for using an instance manager. Need to finish a few things and handle some cases but this captures the main idea. I left some comments on my intentions for some WIP stuff, I am mostly concerned about the InstanceRecord class. Not sure if it is a good use of the architecture, was hoping it would allow us to extend the instance support to the cache. Let me know what you think!

File: Editor/Scripts/MeshSyncServerInspector.cs [hunk 15]
```diff
@@ -85,6 +87,17 @@ public void DrawServerSettings(MeshSyncServer t)
         }
     }
 
+    private void DrawInstanceSettings(MeshSyncServer t)
```
Comment (sindharta) [inline Editor/Scripts/MeshSyncServerInspector.cs]: Doesn't seem to draw anything new ?
Perhaps we can put this in another PR that adds new GUI ?

File: Runtime/Scripts/Settings/Data/MeshSyncServerConfig.cs [hunk 16]
```diff
@@ -6,7 +6,9 @@ namespace Unity.MeshSync {
 internal class MeshSyncServerConfig : MeshSyncPlayerConfig {
     internal MeshSyncServerConfig() { }
 
-    internal MeshSyncServerConfig(MeshSyncServerConfig other) : base(other) { }
+    internal MeshSyncServerConfig(MeshSyncServerConfig other) : base(other)
```
Comment (sindharta) [inline Runtime/Scripts/Settings/Data/MeshSyncServerConfig.cs]: There seems to be no change in this file ?

File: Editor/Scripts/MeshSyncEditorConstants.cs [hunk 17]
```diff
@@ -22,7 +22,6 @@ internal static class MeshSyncEditorConstants {
     
     internal static readonly string[] ANIMATION_INTERPOLATION_ENUMS = System.Enum.GetNames( typeof( InterpolationMode ) );
     internal static readonly string[] Z_UP_CORRECTION_ENUMS         = System.Enum.GetNames( typeof( ZUpCorrectionMode ) );
```
Comment (sindharta) [inline Editor/Scripts/MeshSyncEditorConstants.cs]: There seems to be no change in this file ?

File: Runtime/Scripts/SceneCache/SceneCachePlayer.cs [hunk 18]
```diff
@@ -554,7 +554,6 @@ internal enum TimeUnit {
         }
 #endif
         
-        
         if (!m_sceneCache)
```
Comment (sindharta) [inline Runtime/Scripts/SceneCache/SceneCachePlayer.cs]: There seems to be no change in this file ?

File: Runtime/Scripts/SceneCache/SceneCachePlayableAsset.cs [hunk 19]
```diff
@@ -100,7 +100,9 @@ internal class SceneCachePlayableAsset : BaseExtendedClipPlayableAsset<SceneCach
 
         UpdateClipCurve(clip, m_animationCurve);
         m_isSceneCacheCurveExtracted = true;
+#if UNITY_EDITOR
```
Comment (sindharta) [inline Runtime/Scripts/SceneCache/SceneCachePlayableAsset.cs]: Oops, I'll fix and put this in a separate PR.

File: Runtime/Scripts/MeshSyncServer.cs [hunk 20]
```diff
@@ -472,11 +497,24 @@ public void StartServer()
...
     }
 
+    void Start()
+    {
+#if ENABLE_IL2CPP
+        Instance = this;
+#endif
+    }
+
+    void OnDestroy()
```
Comment (sindharta) [inline Runtime/Scripts/MeshSyncServer.cs]: Just calls base, so we can remove this ?

File: Runtime/Scripts/MeshSyncServer.cs [hunk 21]
```diff
@@ -517,6 +555,18 @@ enum ServerVersion {
         INITIAL_0_4_0 = 1, //initial for version 0.4.0-preview 
     
     }
+
+#if UNITY_EDITOR
+    [SerializeField]
+    private bool m_foldInstanceSettings = true;
```
Comment (sindharta) [inline Runtime/Scripts/MeshSyncServer.cs]: Doesn't seem to be used in this PR for drawing new GUI

File: Runtime/Scripts/MeshSyncServer.cs [hunk 22]
```diff
@@ -23,10 +25,19 @@ public class MeshSyncServer : BaseMeshSync {
     public ServerMessageCallback OnPostRecvMessageCallback = null;
     
 //----------------------------------------------------------------------------------------------------------------------
-    
+
+#if ENABLE_IL2CPP
+    // When building with IL2CPP, we must use a static delegate
+    private static MeshSyncServer Instance;
```
Comment (sindharta) [inline Runtime/Scripts/MeshSyncServer.cs]: This looks like a bug fix.
Can you please separate the IL2CPP support changes into a separate PR ?

File: Runtime/Scripts/MeshSyncInstanceRenderer.cs [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+            LightProbeUsage lightProbeUsage, 
+            LightProbeProxyVolume lightProbeProxyVolume)
+        {
+            if (cameras == null)
+                return;
+            
+            for (var i = 0; i < cameras.Length; i++)
+            {
+                var camera = cameras[i];
+                if (camera == null)
```
Comment (sindharta) [inline Runtime/Scripts/MeshSyncInstanceRenderer.cs]: Just a thought. Shouldn't we loop over the cameras first if there is a possibility to exit earlier, rather than looping over subMeshes ?

File: Plugin~/Src/MeshSync/SceneGraph/msInstanceInfo.cpp [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+	InstanceInfo::InstanceInfo(InstanceInfo&& v) noexcept
+	{
+		*this = std::move(v);
+	}
+
+	InstanceInfo& InstanceInfo::operator=(InstanceInfo&& v)
+	{
+		path = std::move(v.path);
+		transforms = std::move(v.transforms);
+		return *this;
```
Comment (sindharta) [inline Plugin~/Src/MeshSync/SceneGraph/msInstanceInfo.cpp:66]: Are you using tabs with 8 spaces ? 
We can probably clean this up in another PR, but I'd just like to check if that's the intent here.

File: Runtime/Scripts/BaseMeshSync.cs [hunk 25]
```diff
@@ -10,7 +10,6 @@
 using System.IO;
```
Comment (sindharta) [inline Runtime/Scripts/BaseMeshSync.cs]: Linux so file hasn't been updated ?
I can help with this if necessary

File: Runtime/Scripts/SceneCache/SceneCachePlayableAsset.cs (hunk 19 above)
Comment (sindharta) [reply Runtime/Scripts/SceneCache/SceneCachePlayableAsset.cs]: https://github.com/unity3d-jp/MeshSync/pull/640

File: Runtime/Scripts/BaseMeshSync.cs [hunk 26]
```diff
@@ -2197,6 +2435,9 @@ protected virtual void OnDisable()
     private readonly           Dictionary<string, EntityRecord> m_clientObjects = new Dictionary<string, EntityRecord>();
     private protected readonly Dictionary<int, EntityRecord>    m_hostObjects   = new Dictionary<int, EntityRecord>();
     private readonly           Dictionary<GameObject, int>      m_objIDTable    = new Dictionary<GameObject, int>();
+    
+    private protected readonly Dictionary<string, InstanceInfoRecord> m_clientInstances =  new Dictionary<string, InstanceInfoRecord>();
```
Comment (sindharta) [inline Runtime/Scripts/BaseMeshSync.cs]: Can you use FilmInternalUtilities.SerializedDictionary to simplify this a bit ?
The existing code should be updated as well, but it will be great if we can simplify new code from the beginning

File: Runtime/Scripts/InstanceRenderingInfo.cs [hunk 27]
```diff
@@ -0,0 +1,238 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Rendering;
+
+namespace Unity.MeshSync
+{
+    public class InstanceRenderingInfo
```
Comment (sindharta) [inline Runtime/Scripts/InstanceRenderingInfo.cs]: Please make this an internal class, unless we want to open this to users.
If that's the case, we'll also need to write the API documentations.

File: Runtime/Scripts/InstanceRenderingInfo.cs [hunk 28]
```diff
@@ -0,0 +1,238 @@
...
+    {
+        public Mesh Mesh;
+        
+        public List<Matrix4x4[]> DividedInstances { get; private set; } = new List<Matrix4x4[]>();
+
+        private Material[] m_materials;
+
+        [SerializeField]
+        public Material[] Materials
```
Comment (sindharta) [inline Runtime/Scripts/InstanceRenderingInfo.cs]: One convention that I'd like to adopt is to use `camelCase` for properties, following the convention used by Unity.
I know that a lot of MeshSync code don't necessarily reflect this at the moment, but it'll be good if we can start doing this.

File: Runtime/Scripts/InstanceRenderingInfo.cs [hunk 29]
```diff
@@ -0,0 +1,238 @@
...
+
+namespace Unity.MeshSync
+{
+    public class InstanceRenderingInfo
+    {
+        public Mesh Mesh;
+        
+        public List<Matrix4x4[]> DividedInstances { get; private set; } = new List<Matrix4x4[]>();
+
+        private Material[] m_materials;
```
Comment (sindharta) [inline Runtime/Scripts/InstanceRenderingInfo.cs:14]: Shouldn't `m_materials` be serialized ?

File: Runtime/Scripts/InstanceRenderingInfo.cs [hunk 30]
```diff
@@ -0,0 +1,238 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Rendering;
+
+namespace Unity.MeshSync
+{
```
Comment (sindharta) [inline Runtime/Scripts/InstanceRenderingInfo.cs:7]: [Serializable] not required ?

File: Runtime/Scripts/InstanceRenderingInfo.cs [hunk 31]
```diff
@@ -0,0 +1,238 @@
...
+                    return;
+                
+                m_gameObject = value;
+                
+                OnGameObjectUpdated();
+            }
+        }
+
+        [SerializeField]
+        public Renderer Renderer;
```
Comment (sindharta) [inline Runtime/Scripts/InstanceRenderingInfo.cs]: this seems that it can be made private

File: Runtime/Scripts/InstanceRenderingInfo.cs [hunk 32]
```diff
@@ -0,0 +1,238 @@
...
+                    array[j] = worldMatrix * Instances[i * maxSize + j];
+                }
+                
+                DividedInstances.Add(array);
+            }
+
+            var remainder = Instances.Length % maxSize;
+            if (remainder > 0)
+            {
+                var array = new Matrix4x4[remainder];
```
Comment (sindharta) [inline Runtime/Scripts/InstanceRenderingInfo.cs]: Can we turn this section into a function ? Seems like it's the same / very similar with the section above

File: Runtime/Scripts/BaseMeshSync.cs [hunk 33]
```diff
@@ -2119,7 +2350,7 @@ internal List<AnimationClip> GetAnimationClips()
     }
 #endif
 
-    void OnDestroy() {
+    protected void OnDestroy() {
```
Comment (sindharta) [inline Runtime/Scripts/BaseMeshSync.cs]: why do we need `protected` here ?

File: Runtime/Scripts/BaseMeshSync.cs [hunk 34]
```diff
@@ -1764,6 +1960,40 @@ void UpdateConstraint(ConstraintData data)
...
+             DestroyImmediate(rec.renderer);
+             
+            if (onDeleteInstanceInfo != null)
+                onDeleteInstanceInfo(identifier.name);
+        }
+
+        return ret;
+    }
+
+    internal bool EraseInstancedEntityRecord(Identifier identifier)
```
Comment (sindharta) [inline Runtime/Scripts/BaseMeshSync.cs:1945]: Very similar to `EraseInstanceInfoRecord`.
Perhaps we can
1. use generics to create a new function
2. pass the onDelete function to it.

to encapsulate all the common code ?

File: Runtime/Scripts/BaseMeshSync.cs [hunk 35]
```diff
@@ -1091,32 +1181,41 @@ private protected void UpdateMaterialAssetByDefault(MaterialData src, ModelImpor
             smr.sharedMesh = rec.mesh;
 
             // update bones
-            if (dflags.hasBones) {
+            if (dflags.hasBones)
```
Comment (sindharta) [inline Runtime/Scripts/BaseMeshSync.cs]: We probably should start discussing about the convention, first about the brackets.
Perhaps we can talk about this a bit in our weekly sync meeting ?

That aside, if possible, the PR to change the convention should be separated from the PR to add functionality

File: Runtime/Scripts/BaseMeshSync.cs (hunk 25 above)
Comment (seandillon92) [reply Runtime/Scripts/BaseMeshSync.cs]: I was thinking it might be better we do this after the code reviews since the code might be updated, will leave this comment unresolved until I do so

File: Runtime/Scripts/InstanceRenderingInfo.cs (hunk 29 above)
Comment (seandillon92) [reply Runtime/Scripts/InstanceRenderingInfo.cs:14]: There was an iteration where InstanceRenderingInfo was serializable but we have moved away from that

File: Runtime/Scripts/InstanceRenderingInfo.cs (hunk 30 above)
Comment (seandillon92) [reply Runtime/Scripts/InstanceRenderingInfo.cs:7]: There was an iteration where InstanceRenderingInfo was serializable but we have moved away from that

File: Runtime/Scripts/BaseMeshSync.cs (hunk 33 above)
Comment (seandillon92) [reply Runtime/Scripts/BaseMeshSync.cs]: It was for overriding it on the server, will remove it

File: Runtime/Scripts/BaseMeshSync.cs (hunk 35 above)
Comment (seandillon92) [reply Runtime/Scripts/BaseMeshSync.cs]: Not sure why this was reformatted in the first place but I have reverted it back to the original convention. I agree lets discuss and lets keep it in a separate PR

File: Runtime/Scripts/InstanceRenderingInfo.cs (hunk 28 above)
Comment (seandillon92) [reply Runtime/Scripts/InstanceRenderingInfo.cs]: I agree, I have now updated my files with the Unity convention

File: Runtime/Scripts/BaseMeshSync.cs (hunk 26 above)
Comment (seandillon92) [reply Runtime/Scripts/BaseMeshSync.cs]: Done now. For some reason using the SerializedDictionary with template arguments did not work. It was not receiving the serialization callbacks, and when I invoked the serialization events of the dictionary from the handler in the BaseMeshSync class, the keys and values arrays did not persist. I was using the SerializeField attribute. I solved it by creating classes that inherit from SerializedDictionary: EntityRecordDictionary and InstanceInfoRecordDictionary. Is there a better way to solve this?

File: Plugin~/Src/MeshSync/SceneGraph/msInstanceInfo.cpp (hunk 24 above)
Comment (seandillon92) [reply Plugin~/Src/MeshSync/SceneGraph/msInstanceInfo.cpp:66]: I am using 4 spaces tab, yes it would be good to fix it along with all the styling stuff. Not sure why it renders with 8 spaces here

File: Runtime/Scripts/BaseMeshSync.cs (hunk 26 above)
Comment (sindharta) [reply Runtime/Scripts/BaseMeshSync.cs]: > SerializedDictionary with template arguments

How did you declare the structure ?
Indeed though, Unity 2019.x lacks some support in terms of serialization compared to Unity 2020.x and above.
See `DummySerializedDictionaryComponent` test file as a reference.

Comment (seandillon92) [reply Runtime/Scripts/BaseMeshSync.cs]: My declaration was
`[SerializeField] private SerializedDictionary<string, InstanceInfoRecord> m_clientInstances = new SerializedDictionary<string, InstanceInfoRecord>` 
Had a look, my approach matches what I see in that file

File: Runtime/Scripts/BaseMeshSync.cs (hunk 25 above)
Comment (seandillon92) [reply Runtime/Scripts/BaseMeshSync.cs]: I have now updated mscore for all three platforms

Comment (seandillon92) [reply Runtime/Scripts/BaseMeshSync.cs]: @sindharta Jobs are failing as they cannot find the mscore file, so I need your help there. I was building for x86_64 on Ubuntu, any chance the tests are using a different architecture?

**Comment by seandillon92** [comment]
Moving to this branch to avoid issue with CLA-client https://github.com/unity3d-jp/MeshSync/pull/642
